Fix winrm handling of early connection termination or timeout

pecigonzalo opened the pull request:
PR https://github.com/masterzen/winrm/pull/25 introduced a bug where winrm will exit with error code 1 even when err is empty and a lot of error codes from further on the line are omitted (shell.Close()).
This is affecting remote commands as ```net stop winrm``` or ```Stop-Service winrm``` needed when rebooting a host to ensure winrm does not reconnect too soon on a subsecuent command while rebooting.

This PR is not complete yet, the intention is to get some help on this code as i add to it, since go is not my primary lang.

Reference:
https://github.com/mitchellh/packer/pull/2243
https://github.com/packer-community/packer-windows-plugins/issues/54

pecigonzalo commented:
To add more information to this. The slurpAllOutput() function does not check for the connection to be open on any way before trying to get the output as it assumes the connection is open, so if the connection was dropped by the command executed or something on the remote node it gives a EOF error.

pecigonzalo commented:
@masterzen Could you give this a look and let me know if im missing something or something is not following your guidelines?

thanks!

pecigonzalo commented:
Also, the build i failing but i dont understand why, it compiles and work fine on 3 computers, different OSs... is the builder broken?

masterzen commented:
@pecigonzalo the failure happens on Go 1.2, which might be the reason why you're not seeing the issue when compiling locally.

masterzen commented:
Thanks for your work!

It isn't mergeable for the moment, but I'm pretty sure it will be merged when the various minor comments I have will be fixed.

Now, some comments:

* your commits don't seem to come from the same author, can you check your git settings?
* some commits needs squashing (see individual comments)
* it would be great to reproduce the error on Go 1.2 and understand the issue (and fix it)
* I would appreciate a more descriptive commit message so that we can later on understand what we were trying to fix at that time (I had to read the referenced tickets to understand what we was the use case).

Let me know if I can help,
Thanks!
Brice

pecigonzalo commented:
Thanks for the reply, ill work on those fixes, ill ask you about the particular things i dont know or need more info!

pecigonzalo commented:
@masterzen I fixed most of the issues you mentioned i believe and reworked some of the things. I still cant figure out the go 1.2 stuff ill setup an environment for that tonight and commit the fixes.

Feel free to comment on the code again, as i said, golang is not my primary lang.

masterzen commented:
@pecigonzalo thanks for the new commits, I haven't yet had the time to review those. I'll try to do that in the course of the day if I can. At first glance the code looks ok (and golang is really not my primary language too :)

Regarding the problem on go 1.2, we have the exact same with go 1.3 (and it also fails with the new batch of commits you did).

It looks like one of the test is timing out, which forces the go framework to abort the test process. Are you able to run the test correctly on your machine?

Since this patch fixes an incorrect behavior of the library, I would really really appreciate if you could add some tests covering those use cases, otherwise I'm afraid a subsequent patch will remove those new behavior without any of us noticing.

pecigonzalo commented:
Agreed, will add the tests to cover the fixes ASAP. 
Thanks again.

masterzen commented:
@pecigonzalo I've reproduced the test hanging locally with go 1.2. The tests pass with go 1.4 and 1.3.
It's blocking indefinitely here (https://github.com/masterzen/winrm/blob/master/winrm/command_test.go#L82).

Note that this doesn't happen on master, only on your branch, so that's one of the patch that makes this happen.

When I ran git bisect, it isolated: https://github.com/pecigonzalo/winrm/commit/c4a2499cb4449ff63d9b091f4f8e84988e1ba958 as the culprit.

masterzen commented:
@pecigonzalo I stand corrected it's commit https://github.com/pecigonzalo/winrm/commit/3501d5e60cd146cfed89fe3142633c575414d833 that starts failing.

You can reproduce the problem with go 1.2 and using this test.sh file:
```sh
#!/bin/sh
make clean
timeout 30s make test
```
and the following git bisect.
```sh
git bisect start HEAD 132339029dfa67fd39ff8edeed2af78f2cca4fbb
git bisect run /bin/bash test.sh
git bisect log
```

pecigonzalo commented:
Thanks for that, ill take a look at it.

pecigonzalo commented:
FYI, still working on this, just struggling to find time.

masterzen commented:
@pecigonzalo no problem, I also don't have a lot of time hacking on this project. I will try though to understand the hang issue in the tests during the next week-end, if that can help.

masterzen commented:
@pecigonzalo,
I've finally found the issue with the tests hanging on go 1.2. It wasn't coming from your patch, and we apparently had this issue for some times, but since the problem didn't quite appear as frequently as of now, it hadn't been noticed.
I've fixed the problem on master, so you might want to rebase this PR on top of it.
I'll merge your branch as soon as you'll have time to add some tests about the behavior change.
Let me know if you need help for the tests.

Thanks!

pecigonzalo commented:
That is amazing, thanks for the hard work on that! ill rebase as soon as i can but before we marge in addition to the tests ill like to add some other work i have in mind related to this.

Again great stuff!

pecigonzalo commented:
Re-base done. Ill continue the work ASAP.

pecigonzalo commented:
@masterzen Reworked some of the functionality to what i believe is a better overall result. Added some tests, i would like some help on other tests that we need here, and some review of the ones i already created.

Thanks

masterzen commented:
I will have a look ASAP.

hoenirvili commented:
@masterzen close or merge?

masterzen commented:
I'd like to merge it, but it will require some work to make current, and check if it is still usable.

masterzen commented:
Closing in favor of PR #50.
